common/string_calls.c

```diff
@@ -775,3 +776,50 @@ g_strtrim(char *str, int trim_flags)
     return 0;
 }
 
+/*****************************************************************************/
+char *
+g_strnjoin(char *dest, int dest_len, const char *joiner, const char *src[], int src_len)
+{
+    int len = 0;
+    int joiner_len;
+    int i = 0;
+    int dest_remaining;
+    char *dest_pos = dest;
+    char *dest_end;
+    
+    if (dest == NULL || dest_len < 1)
+    {
+        return dest;
+    }
+    if (src == NULL || src_len < 1)
+    {
+        dest[0] = '\0';
+        return dest;
+    }
+
+    dest[0] = '\0';
+    dest_end = dest + dest_len - 1;
+    joiner_len = g_strlen(joiner);
+    for (i = 0; i < src_len - 1 && dest_pos < dest_end; i++)
+    {
+        len = g_strlen(src[i]);
+        dest_remaining = dest_end - dest_pos;
+        g_strncat(dest_pos, src[i], dest_remaining);
+        dest_pos += MIN(len, dest_remaining);
+
+        if (dest_pos < dest_end)
+        {
+            dest_remaining = dest_end - dest_pos;
+            g_strncat(dest_pos, joiner, dest_remaining);
+            dest_pos += MIN(joiner_len, dest_remaining);
+        }
+    }
+
+    if (i == src_len - 1 && dest_pos < dest_end)
+    {
+        g_strncat(dest_pos, src[i], dest_end - dest_pos);
+    }
+
+    dest[dest_len] = '\0';
```

Comment:
Off by one - Should be `dest[dest_len - 1] = '\0`, or `*dest_end = '\0';`

This test catches it:-

```c
START_TEST(test_strnjoin__doesnt_write_off_end)
{
    /* setup */

    const char *src[] = { "a","b","c"};
    char result[5+1+1]; /* a-b-c + term + guard value */

    /* test */

    result[6] = '\x7f';
    g_strnjoin(result, 5 + 1, "-", src, 3);

    /* verify */

    ck_assert_int_eq(result[6], '\x7f');
}
END_TEST
```
Alternatively remove the whole line. You're using `strncat()` so this terminator is guaranteed to be present anyway.

BTW - nice use of MIN().

Reply:
you are right on both points. I added the test case you wrote, and removed the unnecessary writing null after the end of the destination buffer.

I had added the null termination because I misread `strncpy` for `strncat` in the following article:

>An entirely different form of string truncation is one that can result from calls to strncpy. Unlike functions such as snprintf and strncat that always append a terminating NUL, when the source string passed to strncpy is longer than the size specified by the third argument, the function truncates the copy without appending a NUL to the end. 
From [Forming Non-NUL-Terminated Sequences](https://developers.redhat.com/blog/2018/05/24/detecting-string-truncation-with-gcc-8/)